File: compiler/src/main/java/io/github/landerlyoung/jenny/CppGlueCodeGenerator.java
/**
 * Copyright 2016 landerlyoung@gmail.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.landerlyoung.jenny;

import java.io.IOException;
import java.io.Writer;
import java.util.LinkedList;
import java.util.List;

import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.element.VariableElement;
import javax.tools.FileObject;
import javax.tools.StandardLocation;

import io.github.landerlyoung.jenny.template.FileTemplate;

/**
 * Author: landerlyoung@gmail.com
 * Date:   2014-12-17
 * Time:   16:03
 * Life with passion. Code with creativity!
 */
public class CppGlueCodeGenerator extends AbsCodeGenerator {
    //header file name
    private String mHeaderName;
    //source file Name
    private String mSourceName;
    private List<Element> mMethods;
    private final NativeClass mNativeClassAnnotation;

    //DONE HandyHelper.toJNIType throwable
    //DONE Use NativeClass to mark generate, NativeCode to add implements
    //DELETE different constant value for different arch
    //DONE fix get package name
    //DONE support for inner class
    //XXXX support for pure c code
    //DONE file output
    //GOING use file template

    public CppGlueCodeGenerator(Environment env, TypeElement clazz) {
        super(env, clazz);
        mMethods = new LinkedList<>();
        NativeClass annotation = clazz.getAnnotation(NativeClass.class);
        if (annotation == null) {
            annotation = AnnotationResolver.INSTANCE.getDefaultImplementation(NativeClass.class);
        }
        mNativeClassAnnotation = annotation;
    }

    @Override
    public void doGenerate() {
        if (init() && !mMethods.isEmpty()) {
            generateHeader();
            generateSource();
        }
    }

    private boolean init() {
        if (!getMClazz().getKind().equals(ElementKind.CLASS)) return false;

        mHeaderName = getCppClassName() + ".h";
        mSourceName = getCppClassName() + ".cpp";
        log("jenny begin generate glue code for class [" + getMClassName() + "]");
        log("header : [" + mHeaderName + "]");
        log("source : [" + mSourceName + "]");

        findNativeMethods();

        return true;
    }

    private String getCppClassName() {
        String fileName = mNativeClassAnnotation.fileName();
        if (fileName.length() > 0) {
            return fileName;
        } else {
            return mNativeClassAnnotation.simpleName() ? getMSimpleClassName() : getMJNIClassName();
        }
    }

    private void findNativeMethods() {
        getMClazz().getEnclosedElements()
              .stream()
              .filter(e -> e.getKind() == ElementKind.METHOD)
              .forEach(e -> {
                  if (e.getModifiers().contains(Modifier.NATIVE)) {
                      mMethods.add(e);
                  } else if (e.getAnnotation(NativeCode.class) != null) {
                      error("Annotation @" + NativeCode.class.getSimpleName()
                              + " should only be applied to NATIVE method!");
                  }
              });
    }

    public void generateHeader() {
        Writer w = null;
        try {
            FileObject fileObject = getMEnv().getFiler().createResource(StandardLocation.SOURCE_OUTPUT, AbsCodeGenerator.PKG_NAME, mHeaderName);
            log("write header file [" + fileObject.getName() + "]");
            w = fileObject.openWriter();


            w.write(FileTemplate.withType(
                    mNativeClassAnnotation.dynamicRegisterJniMethods()
                            ? FileTemplate.Type.JNI_HEADER_SKELETON_NS
                            : FileTemplate.Type.JNI_HEADER_SKELETON)
                                .add("namespace", getCppClassName())
                                .add("full_slash_class_name", getMSlashClassName())
                                .add("full_java_class_name", getMClassName())
                                .add("consts", generateConstantsDefinition())
                                .add("class_prefix", !getMJNIClassName().isEmpty() ? getMJNIClassName() + "_" : "")
                                .add("methods", generateFunctions(false))
                                //ns version only
                                .add("registers", mNativeClassAnnotation.dynamicRegisterJniMethods()
                                        ? FileTemplate.withType(FileTemplate.Type.REGISTER_NATIVE_FUNCTIONS_NS).create()
                                        : "")
                                .create()
            );

            w.close();
        } catch (IOException e) {
            warn("generate header file " + mHeaderName + " failed!");
        } finally {
            IOUtils.closeSilently(w);
        }
    }

    public void generateSource() {
        Writer w = null;
        try {
            FileObject fileObject = getMEnv().getFiler().createResource(StandardLocation.SOURCE_OUTPUT, AbsCodeGenerator.PKG_NAME, mSourceName);
            log("write source file [" + fileObject.getName() + "]");
            w = fileObject.openWriter();

            w.write(FileTemplate.withType(
                    mNativeClassAnnotation.dynamicRegisterJniMethods()
                            ? FileTemplate.Type.NATIVE_CPP_SKELETON_NS
                            : FileTemplate.Type.NATIVE_CPP_SKELETON)
                                .add("header", mHeaderName)
                                .add("namespace", getCppClassName())
                                .add("android_log_marcos", mNativeClassAnnotation.androidLog()
                                        ? FileTemplate.withType(FileTemplate.Type.ANDROID_LOG_MARCOS).create()
                                        : "")
                                .add("full_java_class_name", getMClassName())
                                .add("full_slash_class_name", getMSlashClassName())
                                .add("full_native_class_name", getMJNIClassName())
                                .add("simple_class_name", getMSimpleClassName())
                                .add("methods", generateFunctions(true))
                                .add("dynamic_jni_register", mNativeClassAnnotation.dynamicRegisterJniMethods()
                                        ? FileTemplate.withType(FileTemplate.Type.CPP_DYNAMIC_JNI_REGISTER_NS)
                                                      .add("jni_method_struct", generateJniNativeMethodStruct())
                                                      .add("slash_class_name", getMSlashClassName())
                                                      .create()
                                        : "")
                                .add("jni_onload_impl",
                                        mNativeClassAnnotation.dynamicRegisterJniMethods()
                                                ? FileTemplate.withType(
                                                FileTemplate.Type.JNI_ONLOAD_IMPL_NS)
                                                              .add("namespace", getCppClassName())
                                                              .create()
                                                : "")
                                .create()
            );
        } catch (IOException e) {
            warn("generate source file " + mSourceName + " failed");
        } finally {
            IOUtils.closeSilently(w);
        }
    }

    private String generateConstantsDefinition() {
        StringBuilder sb = new StringBuilder();
        //if this field is a compile-time constant value it's
        //value will be returned, otherwise null will be returned.
        final String constant_prefix = (!mNativeClassAnnotation.dynamicRegisterJniMethods() && !getMJNIClassName().isEmpty() ? getMJNIClassName() + "_" : "");
        getMClazz().getEnclosedElements()
              .stream()
              .filter(e -> e.getKind() == ElementKind.FIELD)
              .map(e -> (VariableElement) e)
              .filter(ve -> ve.getConstantValue() != null)
              .forEach(ve -> {
                  //if this field is a compile-time constant value it's
                  //value will be returned, otherwise null will be returned.
                  Object constValue = ve.getConstantValue();

                  String nativeType = getMHelper().toNativeType(ve.asType(), true);

                  sb.append(FileTemplate.withType(FileTemplate.Type.CONSTANT_TEMPLATE)
                                        .add("type", nativeType)
                                        .add("_const", nativeType.contains("*") ? "const " : "")
                                        .add("name", constant_prefix + ve.getSimpleName().toString())
                                        .add("full_class_name", getMSlashClassName())
                                        .add("value", HandyHelper.getJNIHeaderConstantValue(constValue))
                                        .create()
                  );
              });
        return sb.toString();
    }

    private String generateJniNativeMethodStruct() {
        StringBuilder sb = new StringBuilder();
        int methodLen = mMethods.size();
        for (int i = 0; i < methodLen; i++) {
            ExecutableElement m = (ExecutableElement) mMethods.get(i);
            sb.append(FileTemplate.withType(FileTemplate.Type.NATIVE_JNI_NATIVE_METHOD_STRUCT)
                                  .add("method_name", m.getSimpleName().toString())
                                  .add("signature", getMHelper().getBinaryMethodSignature(m))
                                  .add("comma", i != methodLen - 1 ? "," : "")
                                  .create()
            );
        }
        return sb.toString();
    }

    private String generateFunctions(boolean isSource) {
        StringBuilder sb = new StringBuilder();
        for (Element m : mMethods) {
            ExecutableElement e = (ExecutableElement) m;
            String result =
                    FileTemplate.withType(FileTemplate.Type.NATIVE_METHOD_DECLARE_TEMPLATE)
                                .add("full_java_class_name", getMJNIClassName())
                                .add("modifiers", getMHelper().getMethodModifiers(e))
                                .add("export_", mNativeClassAnnotation.dynamicRegisterJniMethods() ? "" : "JNIEXPORT ")
                                .add("java_return_type", e.getReturnType().toString())
                                .add("note_method_name", e.getSimpleName().toString())
                                .add("method_name", getMethodName(e))
                                .add("java_parameters", getMHelper().getJavaMethodParam(e))
                                .add("jni_return_type", getMHelper().toJNIType(e.getReturnType()))
                                .add("method_signature", getMHelper().getMethodSignature(e))
                                .add("native_parameters", getMHelper().getNativeMethodParam(e))
                                .add("end", isSource ? "" : ";\n\n")
                                .create();
            if (isSource) {
                result = generateFunctionWithReturnStatement(result, e);
            }
            sb.append(result).append("\n");
        }
        if (!isSource && sb.length() != 0 && !mNativeClassAnnotation.dynamicRegisterJniMethods()) {
            sb.insert(0, FileTemplate.withType(FileTemplate.Type.CPP_EXPORT_MARCO_BEGIN).create());
            sb.append(FileTemplate.withType(FileTemplate.Type.CPP_EXPORT_MARCO_END).create());
        }
        return sb.toString();
    }

    private String getMethodName(ExecutableElement m) {
        String simpleName = m.getSimpleName().toString();
        if (mNativeClassAnnotation.dynamicRegisterJniMethods()) {
            return simpleName;
        } else {
            return "Java_" + getMJNIClassName() + "_" + simpleName;
        }
    }

    private String generateFunctionWithReturnStatement(String declare, ExecutableElement m) {
        FileTemplate template = FileTemplate.withType(FileTemplate.Type.NATIVE_METHOD_TEMPLATE)
                                            .add("native_method_declare_template", declare);

        StringBuilder returnStatement = new StringBuilder();
        NativeCode a = m.getAnnotation(NativeCode.class);
        if (a != null) {
            for (String s : a.value()) {
                returnStatement.append("    ")
                               .append(s)
                               .append("\n");
            }
            if (returnStatement.length() > 0) {
                returnStatement.replace(returnStatement.length() - 1, returnStatement.length(), "");
            }
        } else {
            returnStatement.append(getMHelper().getReturnStatement(m));
        }
        return template.add("default_return_statement", returnStatement.toString())
                       .create();
    }
}

File: compiler/src/main/java/io/github/landerlyoung/jenny/CppGlueCodeGenerator.kt
/**
 * Copyright 2016 landerlyoung@gmail.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.landerlyoung.jenny

import java.io.IOException
import java.util.*
import javax.lang.model.element.*
import javax.tools.StandardLocation

/**
 * Author: landerlyoung@gmail.com
 * Date:   2014-12-17
 * Time:   16:03
 * Life with passion. Code with creativity!
 */

//DONE HandyHelper.toJNIType throwable
//DONE Use NativeClass to mark generate, NativeCode to add implements
//DELETE different constant value for different arch
//DONE fix get package name
//DONE support for inner class
//XXXX support for pure c code
//DONE file output
//GOING use file template
class CppGlueCodeGenerator(env: Environment, clazz: TypeElement) : AbsCodeGenerator(env, clazz) {
    private lateinit var mNamespaceHelper: NamespaceHelper
    // header file name
    private lateinit var mHeaderName: String
    // source file Name
    private lateinit var mSourceName: String
    private val mMethods: MutableList<Element> = LinkedList()
    private val mNativeClassAnnotation: NativeClass =
            clazz.getAnnotation(NativeClass::class.java)
                    ?: AnnotationResolver.getDefaultImplementation(NativeClass::class.java)

    private val cppClassName: String = mSimpleClassName

    override fun doGenerate() {
        if (init() && mMethods.isNotEmpty()) {
            generateHeader()
            generateSource()
        }
    }

    private fun init(): Boolean {
        if (mClazz.kind != ElementKind.CLASS) return false

        mNamespaceHelper = NamespaceHelper(mNativeClassAnnotation.namespace)

        mHeaderName = mNamespaceHelper.fileNamePrefix + "$cppClassName.h"
        mSourceName = mNamespaceHelper.fileNamePrefix + "$cppClassName.cpp"
        log("jenny begin generate glue code for class [$mClassName]")
        log("header : [$mHeaderName]")
        log("source : [$mSourceName]")

        findNativeMethods()

        return true
    }

    private fun findNativeMethods() {
        mClazz.enclosedElements
                .asSequence()
                .filter { it.kind == ElementKind.METHOD }
                .forEach { e ->
                    if (e.modifiers.contains(Modifier.NATIVE)) {
                        mMethods.add(e)
                    } else if (e.getAnnotation(NativeCode::class.java) != null) {
                        error("Annotation @" + NativeCode::class.java.simpleName
                                + " should only be applied to NATIVE method!")
                    }
                }
    }

    private fun generateHeader() {
        val fileObject = mEnv.filer.createResource(StandardLocation.SOURCE_OUTPUT, Constants.JENNY_GEN_DIR_GLUE_HEADER, mHeaderName)
        log("write header file [" + fileObject.name + "]")

        fileObject.openOutputStream().use { out ->
            try {
                buildString {
                    append(Constants.AUTO_GENERATE_NOTICE)
                    append("""
                        |
                        |/* C++ header file for class $mSlashClassName */
                        |#pragma once
                        |
                        |#include <jni.h>
                        |
                        |${mNamespaceHelper.beginNamespace()}
                        |namespace $cppClassName {
                        |
                        |// DO NOT modify
                        |static constexpr auto FULL_CLASS_NAME = "$mSlashClassName";
                        |
                        |""".trimMargin())

                    if (mNativeClassAnnotation.dynamicRegisterJniMethods) {
                        buildConstantsDefinition()
                        buildMethodsDefinition(false)
                        buildJniRegister()
                        endNamespace()
                    } else {
                        buildConstantsDefinition()
                        endNamespace()
                        buildMethodsDefinition(false)
                    }
                }.let { content ->
                    out.write(content.toByteArray(Charsets.UTF_8))
                }
            } catch (e: IOException) {
                warn("generate header file $mHeaderName failed!")
            }
        }
    }

    private fun generateSource() {
        val fileObject = mEnv.filer.createResource(StandardLocation.SOURCE_OUTPUT, Constants.JENNY_GEN_DIR_GLUE_SOURCE, mSourceName)
        log("write source file [" + fileObject.name + "]")
        try {
            buildString {
                append(Constants.AUTO_GENERATE_SOURE_NOTICE)
                append("""
                    |#include "$mHeaderName"
                    |
                    |""".trimMargin())

                if (mNativeClassAnnotation.dynamicRegisterJniMethods) {
                    append("""
                        |
                        |${mNamespaceHelper.beginNamespace()}
                        |
                        |""".trimMargin())
                }

                buildMethodsDefinition(true)

                if (mNativeClassAnnotation.dynamicRegisterJniMethods) {
                    endNamespace(true)
                }
            }.let { content ->
                fileObject.openOutputStream().write(content.toByteArray(Charsets.UTF_8))
            }
        } catch (e: IOException) {
            warn("generate source file $mSourceName failed")
        }
    }

    private fun StringBuilder.endNamespace(isSource: Boolean = false) {
        append('\n')
        if (!isSource) {
            append("} // endof namespace $cppClassName\n")
        }
        append("${mNamespaceHelper.endNamespace()}\n\n")
    }

    private fun StringBuilder.buildConstantsDefinition() {
        mClazz.enclosedElements
                .asSequence()
                .filter { it.kind == ElementKind.FIELD }
                .map { it as VariableElement }
                .filter { it.constantValue != null }
                .forEach {
                    // if this field is a compile-time constant value it's
                    // value will be returned, otherwise null will be returned.
                    val constValue = it.constantValue!!

                    val nativeType = mHelper.toNativeType(it.asType(), true)

                    append("static constexpr $nativeType ${it.simpleName} = ${HandyHelper.getJNIHeaderConstantValue(constValue)};\n")
                }
        append('\n')
    }

    private fun StringBuilder.buildMethodsDefinition(isSource: Boolean) {
        val externC = !isSource && !mNativeClassAnnotation.dynamicRegisterJniMethods
        if (externC) {
            append("""
                |
                |#ifdef __cplusplus
                |extern "C" {
                |#endif
                |
                |""".trimMargin())
        }
        mMethods.forEach { m ->
            val e = m as ExecutableElement
            val javaModifiers = mHelper.getModifiers(e)
            val javaReturnType = e.returnType.toString()
            val javaMethodName = e.simpleName.toString()
            val javaParameters = mHelper.getJavaMethodParam(e)
            val javaMethodSignature = mHelper.getMethodSignature(e)
            val export = if (isSource || mNativeClassAnnotation.dynamicRegisterJniMethods) "" else "JNIEXPORT "
            val jniCall = if (isSource) "" else "JNICALL "
            val jniReturnType = mHelper.toJNIType(e.returnType)
            val nativeMethodName =
                    if (isSource && mNativeClassAnnotation.dynamicRegisterJniMethods)
                        cppClassName + "::" + getMethodName(e)
                    else
                        getMethodName(e)
            val nativeParameters = mHelper.getNativeMethodParam(e)

            append("""
            |/*
            | * Class:     $mJNIClassName
            | * Method:    $javaModifiers $javaReturnType ${javaMethodName}(${javaParameters})
            | * Signature: $javaMethodSignature
            | */
            |${export}${jniReturnType} ${jniCall}${nativeMethodName}(${nativeParameters})""".trimMargin())

            if (isSource) {
                buildMethodBodyWithReturnStatement(m)
            } else {
                append(';')
            }
            append("\n\n")
        }

        if (externC) {
            append("""|#ifdef __cplusplus
                      |}
                      |#endif""".trimMargin())
        }
    }

    private fun StringBuilder.buildMethodBodyWithReturnStatement(m: ExecutableElement) {
        append(" {\n")
        val a = m.getAnnotation(NativeCode::class.java)
        if (a != null) {
            for (line in a.value) {
                append("    ")
                append(line)
                append('\n')
            }
        } else {
            append("    ")
            append(mHelper.getReturnStatement(m))
            append('\n')
        }
        append("}")
    }

    private fun StringBuilder.buildJniRegister() {
        append("""
            |/**
            |* register Native functions
            |* @returns success or not
            |*/
            |inline bool registerNativeFunctions(JNIEnv *env) {
            |   const JNINativeMethod gsNativeMethods[] = {
            |""".trimMargin())
        buildJniNativeMethodStructs()
        append("""
            |   };
            |   const int gsMethodCount =
            |       sizeof(gsNativeMethods) / sizeof(JNINativeMethod);
            |
            |   bool success = false;
            |   jclass clazz = env->FindClass(FULL_CLASS_NAME);
            |   if (clazz != nullptr) {
            |       success = 0 == env->RegisterNatives(clazz, gsNativeMethods, gsMethodCount);
            |       env->DeleteLocalRef(clazz);
            |   }
            |   return success;
            |}
            |""".trimMargin())
    }

    private fun StringBuilder.buildJniNativeMethodStructs() {
        val it = mMethods.iterator()
        while (it.hasNext()) {
            val m = it.next() as ExecutableElement
            val methodName = m.simpleName.toString()
            val signature = mHelper.getBinaryMethodSignature(m)
            append("""
            |       {
            |           /* method name      */ const_cast<char *>("$methodName"),
            |           /* method signature */ const_cast<char *>("$signature"),
            |           /* function pointer */ reinterpret_cast<void *>(${methodName})
            |       }""".trimMargin())
            if (it.hasNext()) {
                append(",")
            }
            append('\n')

        }
    }

    private fun getMethodName(m: ExecutableElement): String {
        val simpleName = m.simpleName.toString()
        return if (mNativeClassAnnotation.dynamicRegisterJniMethods) {
            simpleName
        } else {
            "Java_" + mJNIClassName + "_" + simpleName
        }
    }

}
